### PR TITLE
fix connecctivity analysis with fip - issue #81

### PR DESCRIPTION
**Pull request by adisos**

### CODE REVIEW

File: pkg/vpcmodel/nodesConnectivity.go [hunk 1]
```diff
@@ -50,63 +53,90 @@ func (c *CloudConfig) GetVPCNetworkConnectivity(grouping bool) *VPCConnectivity
...
-				}
-				perLayerRes[layerName][peerNode] = filteredConns
-				allowedConnsBetweenCapturedAndPeerNode = allowedConnsBetweenCapturedAndPeerNode.Intersection(filteredConns)
-				// do not break if empty, to enable computation for all layers
+			if appliedRouter == nil {
+				// without fip/pgw there is no external connectivity
+				allLayersRes[peerNode] = NoConns()
+				continue
+			}
+			// appliedFilters are either both nacl and sg (for pgw) or only sg (for fip)
```
Comment (zivnevo) [inline pkg/vpcmodel/nodesConnectivity.go:136]: This comment might be IBM-specific. Consider rephrasing.

File: pkg/vpcmodel/nodesConnectivity.go [hunk 2]
```diff
@@ -165,6 +230,12 @@ func (v *VPCConnectivity) computeAllowedStatefulConnections() {
 
 	for src, connsMap := range v.AllowedConnsCombined {
 		for dst, conn := range connsMap {
+			// iterate pairs (src,dst) with conn as allowed connectivity, to check stateful aspect
+			if v.isConnExternalThroughFIP(src, dst) {
+				updateAllowedConnsMap(v.AllowedConnsCombinedStateful, src, dst, conn)
+				continue
+			}
```
Comment (zivnevo) [inline pkg/vpcmodel/nodesConnectivity.go:237]: This might be an IBM-specific thing. For now maybe just add a comment or a TODO.

**Review by zivnevo — CHANGES_REQUESTED** [review]

File: pkg/ibmvpc/examples/acl_testing3_with_grouping.txt [hunk 3]
```diff
@@ -1,8 +1,9 @@
 
 grouped output:
+1.0.0.0/8,100.128.0.0/9,11.0.0.0/8,128.0.0.0/5,142.0.0.0/8,143.0.0.0/8,147.235.219.206/32,147.235.219.207/32,147.235.219.208/28,161.26.0.0/16,161.27.0.0/16,169.255.0.0/16,172.32.0.0/11,192.0.1.0/24,192.0.3.0/24,192.169.0.0/16,192.88.100.0/22,198.20.0.0/14,198.51.101.0/24,203.0.114.0/23 => vsi2-ky[10.240.20.4] : All Connections
 vsi1-ky[10.240.10.4] => 161.26.0.0/16 : protocol: UDP  
 vsi1-ky[10.240.10.4] => vsi2-ky[10.240.20.4] : protocol: TCP,UDP  
-vsi2-ky[10.240.20.4] => 142.0.0.0/8 : protocol: ICMP  
+vsi2-ky[10.240.20.4] => 1.0.0.0/8,100.128.0.0/9,11.0.0.0/8,128.0.0.0/5,142.0.0.0/8,143.0.0.0/8,147.235.219.206/32,147.235.219.207/32,147.235.219.208/28,161.26.0.0/16,161.27.0.0/16,169.255.0.0/16,172.32.0.0/11,192.0.1.0/24,192.0.3.0/24,192.169.0.0/16,192.88.100.0/22,198.20.0.0/14,198.51.101.0/24,203.0.114.0/23 : All Connections
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/ibmvpc/examples/acl_testing3_with_grouping.txt:6]: opened issue #100

File: pkg/ibmvpc/examples/acl_testing3.md [hunk 4]
```diff
@@ -1,9 +1,48 @@
...
+| Public Internet [161.27.0.0/16] | vsi2-ky[10.240.20.4] | All Connections |
+| Public Internet [169.255.0.0/16] | vsi2-ky[10.240.20.4] | All Connections |
+| Public Internet [172.32.0.0/11] | vsi2-ky[10.240.20.4] | All Connections |
+| Public Internet [192.0.1.0/24] | vsi2-ky[10.240.20.4] | All Connections |
+| Public Internet [192.0.3.0/24] | vsi2-ky[10.240.20.4] | All Connections |
+| Public Internet [192.169.0.0/16] | vsi2-ky[10.240.20.4] | All Connections |
+| Public Internet [192.88.100.0/22] | vsi2-ky[10.240.20.4] | All Connections |
+| Public Internet [198.20.0.0/14] | vsi2-ky[10.240.20.4] | All Connections |
+| Public Internet [198.51.101.0/24] | vsi2-ky[10.240.20.4] | All Connections |
+| Public Internet [203.0.114.0/23] | vsi2-ky[10.240.20.4] | All Connections |
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/ibmvpc/examples/acl_testing3.md:23]: there should not be limited access, it seems there is bug in adding external ip-blocks at the parsing stage, opened issue #101

File: pkg/vpcmodel/nodesConnectivity.go [hunk 5]
```diff
@@ -154,6 +184,41 @@ func getConnectionStr(src, dst, conn, suffix string) string {
...
+	return false
+}
+
+func updateAllowedConnsMap(m map[Node]map[Node]*common.ConnectionSet, src, dst Node, conn *common.ConnectionSet) {
+	if _, ok := m[src]; !ok {
+		m[src] = map[Node]*common.ConnectionSet{}
+	}
+	m[src][dst] = conn
+}
+
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/vpcmodel/nodesConnectivity.go:221]: added a new type for `map[Node]map[Node]*common.ConnectionSet` in the next PR

File: pkg/vpcmodel/nodesConnectivity.go [hunk 6]
```diff
@@ -50,63 +53,90 @@ func (c *CloudConfig) GetVPCNetworkConnectivity(grouping bool) *VPCConnectivity
...
 }
 
+func (c *CloudConfig) getFilterTrafficResourceOfKind(kind string) FilterTrafficResource {
+	for _, filter := range c.FilterResources {
+		if filter.Kind() == kind {
+			return filter
+		}
+	}
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/vpcmodel/nodesConnectivity.go]: done

File: pkg/vpcmodel/nodesConnectivity.go [hunk 7]
```diff
@@ -50,63 +53,90 @@ func (c *CloudConfig) GetVPCNetworkConnectivity(grouping bool) *VPCConnectivity
...
+func (c *CloudConfig) getFiltersAllowedConnsBetweenNodesPerDirectionAndLayer(
+	src, dst Node,
+	isIngress bool,
+	layer string) *common.ConnectionSet {
+	filter := c.getFilterTrafficResourceOfKind(layer)
+	if filter == nil {
+		return AllConns()
+	}
+	filteredConns := filter.AllowedConnectivity(src, dst, isIngress)
+	return filteredConns
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/vpcmodel/nodesConnectivity.go]: done

File: pkg/vpcmodel/nodesConnectivity.go (hunk 1 above)
Comment (adisos) [reply pkg/vpcmodel/nodesConnectivity.go]: done

File: pkg/vpcmodel/nodesConnectivity.go (hunk 2 above)
Comment (adisos) [reply pkg/vpcmodel/nodesConnectivity.go]: done

**Review by zivnevo — APPROVED** [review]

**Pull request merged**